### PR TITLE
file based config support for galley

### PR DESCRIPTION
**Pull request by irisdingbj**

### CODE REVIEW

**Comment by irisdingbj** [comment]
@ozevren

File: galley/pkg/fs/fsutilities.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+type fileResourceKey struct {
+	key  string
+	kind string
+}
+
+func (r *istioResource) key() string {
+	if len(r.u.GetNamespace()) > 0 {
+		return r.u.GetNamespace() + "/" + r.u.GetName()
```
Comment (ozevren) [inline galley/pkg/fs/fsutilities.go]: This is prone to creating garbage. Consider calculating it once and storing it.

File: galley/pkg/fs/fssource.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+				if more && ev.IsDelete() {
+					s.deleteResource(ev.Name)
+				}
+				if more && ev.IsCreate() {
+					fi, err := os.Stat(ev.Name)
+					if err != nil {
+						scope.Warnf("error occurs for watching %s", ev.Name)
+					} else {
+						if fi.Mode().IsDir() {
+							scope.Debugf("add wathcher for new folder %s", ev.Name)
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]: typo

File: galley/pkg/fs/fssource.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+	}
+	s.watcher.Watch(s.root)
+	s.initialCheck()
+	go func() {
+		for {
+			select {
+			// watch for events
+			case ev, more := <-s.watcher.Event:
+				if more && ev.IsDelete() {
+					s.deleteResource(ev.Name)
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]: Rename to deleteResourcesInFile?

File: galley/pkg/fs/fssource.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+		return nil, err
+	}
+	s.watcher.Watch(s.root)
+	s.initialCheck()
+	go func() {
+		for {
+			select {
+			// watch for events
+			case ev, more := <-s.watcher.Event:
+				if more && ev.IsDelete() {
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]:
``` if !more {
break
}

if ev.IsDelete() {
} else if ev.IsCreate() {
}

```

File: galley/pkg/fs/fssource.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+		}
+		if len(fileResorceKeys) == 0 {
+			delete(s.fileResorceKeys, fileName)
+		}
+	}
+
+}
+
+func (s *fsSource) pubEvent(fileName string, newData *map[string]*istioResource) {
+	s.mu.Lock()
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]: processAddOrUpdate
processModifications
processUpdates?

File: galley/pkg/fs/fssource.go [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+		result[r.key()] = r
+		resourceKeyList = append(resourceKeyList, &fileResourceKey{r.key(), r.u.GetKind()})
+	}
+	if initial {
+		s.fileResorceKeys[path] = resourceKeyList
+	}
+	return result
+}
+
+func (s *fsSource) checkDeleteResourcesInFile(fileName string, newData *map[string]*istioResource) {
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]: processDeletes?

File: galley/pkg/fs/fssource.go (hunk 5 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: processAddOrUpdate

File: galley/pkg/fs/fsutilities.go (hunk 1 above)
Comment (irisdingbj) [reply galley/pkg/fs/fsutilities.go]: done, move to istioResource

File: galley/pkg/fs/fssource.go [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+var supportedExtensions = map[string]bool{
+	".yaml": true,
+	".yml":  true,
+}
+var scope = log.RegisterScope("file-source", "Source for File System", 0)
+
+// fsSource is source implementation for filesystem.
+type fsSource struct {
+	//Config File Path
```
[the remark this answers is not in the document]
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: done

File: galley/pkg/fs/fssource.go (hunk 6 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: processDelete

File: galley/pkg/fs/fssource.go (hunk 5 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: processAddOrUpdate

File: galley/pkg/fs/fssource.go (hunk 4 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: done

File: galley/pkg/fs/fssource.go (hunk 2 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: done

File: galley/pkg/fs/fssource.go (hunk 3 above)
Comment (irisdingbj) [reply galley/pkg/fs/fssource.go]: processPartialDelete

File: galley/pkg/fs/fssource.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+			delete(s.fileResorceKeys, fileName)
+		}
+	}
+
+}
+
+func (s *fsSource) processAddOrUpdate(fileName string, newData *map[string]*istioResource) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	//need versionUpdated as sometimes when fswatcher fires events, there is actually no change on the file content
```
Comment (ozevren) [inline galley/pkg/fs/fssource.go]: Nit: please leave spaces after "//"

File: galley/pkg/fs/fssource_test.go [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+	s, err := New(fst.rootPath)
+	if err != nil {
+		t.Fatalf("Unexpected error found: %v", err)
+	}
+
+	if s == nil {
+		t.Fatal("expected non-nil source")
+	}
+}
+func TestFsSource_InitialScan(t *testing.T) {
```
Comment (ozevren) [inline galley/pkg/fs/fssource_test.go]: Nit : You can make these tests fully declarative, which allows us to add more tests quickly as needed.

**Comment by ozevren** [comment]
You will need to rebase/merge to the head to get the circle tests to pass.

**Pull request merged**